File: 0020_Valid_Parentheses/main.cpp
// T = O(n), S = O(n)
class Solution {
  public:
    bool isValid(string s) {
        stack<char> st;
        for (auto c : s) {
            if (c == '(') {
                st.push(')');
            } else if (c == '{') {
                st.push('}');
            } else if (c == '[') {
                st.push(']');
            } else {
                if (st.empty() || st.top() != c) {
                    return false;
                }
                st.pop();
            }
        }
        return st.empty();
    }
};

File: 0020_Valid_Parentheses/golang/main.go
// T = O(n), S = O(n)
func isValid(s string) bool {
	if len(s) == 0 {
		return true
	}
	st := []rune{}
	for _, c := range s {
		if c == '(' {
			st = append(st, ')')
		} else if c == '{' {
			st = append(st, '}')
		} else if c == '[' {
			st = append(st, ']')
		} else {
			if len(st) == 0 || st[len(st)-1] != c {
				return false
			}
			st = st[:len(st)-1]
		}
	}
	return len(st) == 0
}